src/Adapter/Product/ProductDuplicator.php

```diff
@@ -0,0 +1,644 @@
+     * @param int $oldProductId
+     *
+     * @throws CannotUpdateProductException
+     */
+    private function updateSearchIndexation(Product $newProduct, int $oldProductId): void
+    {
+        $productIsVisibleInSearch = in_array(
+            $newProduct->visibility,
+            [ProductVisibility::VISIBLE_EVERYWHERE, ProductVisibility::VISIBLE_IN_SEARCH]
+        );
+
+        if (!$this->isSearchIndexationOn || !$productIsVisibleInSearch) {
+            return;
+        }
+
+        try {
+            if (!Search::indexation(false, $newProduct->id)) {
+                throw new CannotUpdateProductException(
+                    sprintf('Cannot update search indexation when duplicating product %d', $oldProductId),
+                    CannotUpdateProductException::FAILED_UPDATE_SEARCH_INDEXATION
+                );
+            }
+        } catch (PrestaShopException $e) {
+            throw new CoreException(
+                sprintf('Error occurred when duplicating product %d. Failed to update search indexation', $oldProductId),
+                0,
+                $e
+            );
+        }
+    }
+
+    /**
+     * @param Product $product
+     *
+     * @return Product the new product
+     */
+    private function duplicateProduct(Product $product): Product
+    {
+        $product->name = $this->getNewProductName($product->name);
+        $this->setPriceByShops($product);
+
+        $product->indexed = false;
+        $product->active = false;
+
+        return $this->productRepository->duplicate($product);
+    }
+
+    /**
+     * Provides duplicated product name
+     *
+     * @param array<int, string> $oldProductLocalizedNames
+     *
+     * @return array<int, string>
+     */
+    private function getNewProductName(array $oldProductLocalizedNames): array
+    {
+        $newProductLocalizedNames = [];
+        foreach ($oldProductLocalizedNames as $langId => $oldName) {
+            $langId = (int) $langId;
+            $namePattern = $this->translator->trans('copy of %s', [], 'Admin.Catalog.Feature', Language::getLocaleById($langId));
+            $newName = sprintf($namePattern, $oldName);
+            $newProductLocalizedNames[$langId] = $this->stringModifier->cutEnd($newName, ProductSettings::MAX_NAME_LENGTH);
+        }
+
+        return $newProductLocalizedNames;
+    }
+
+    /**
+     * @param Product $product
+     */
+    private function setPriceByShops(Product $product): void
+    {
+        if (!empty($product->price) || !$this->multistoreContextChecker->isGroupShopContext()) {
+            return;
+        }
+
+        foreach ($this->getContextShops() as $shop) {
+            $priceByShop = $this->productRepository->getPriceByShop(
+                new ProductId((int) $product->id),
+                new ShopId((int) $shop['id_shop'])
+            );
+
+            if (!$priceByShop) {
+                continue;
+            }
+
+            $product->price = (float) (string) $priceByShop;
```

Comment:
In DecimalNumber, can we have a `toFloat()` method ?

Reply:
I already asked for it but it was not accepted because it's almost an anti-pattern as the purpose of the Decimal lib is not to use float any more We are forced to convert them because of retro-compatibility

Reply:
Correct.